invokeai/app/services/download_manager.py

```diff
@@ -0,0 +1,180 @@
+        """
+        List active DownloadJobBases.
+
+        :returns List[DownloadJobBase]: List of download jobs whose state is not "completed."
+        """
+        pass
+
+    @abstractmethod
+    def id_to_job(self, id: int) -> DownloadJobBase:
+        """
+        Return the DownloadJobBase corresponding to the string ID.
+
+        :param id: ID of the DownloadJobBase.
+
+        Exceptions:
+        * UnknownJobIDException
+        """
+        pass
+
+    @abstractmethod
+    def start_all_jobs(self):
+        """Enqueue all stopped jobs."""
+        pass
+
+    @abstractmethod
+    def pause_all_jobs(self):
+        """Pause and dequeue all active jobs."""
+        pass
+
+    @abstractmethod
+    def cancel_all_jobs(self):
+        """Cancel all active and enquedjobs."""
+        pass
+
+    @abstractmethod
+    def start_job(self, job: DownloadJobBase):
+        """Start the job putting it into ENQUEUED state."""
+        pass
+
+    @abstractmethod
+    def pause_job(self, job: DownloadJobBase):
+        """Pause the job, putting it into PAUSED state."""
+        pass
+
+    @abstractmethod
+    def cancel_job(self, job: DownloadJobBase):
+        """Cancel the job, clearing partial downloads and putting it into ERROR state."""
+        pass
+
+    @abstractmethod
+    def change_priority(self, job: DownloadJobBase, delta: int):
+        """
+        Change the job's priority.
+
+        :param job: Job to  apply change to
+        :param delta: Value to increment or decrement priority.
+
+        Lower values are higher priority.  The default starting value is 10.
+        Thus to make this a really high priority job:
+           job.change_priority(-10).
+        """
+        pass
+
+    @abstractmethod
+    def join(self):
+        """Wait until all jobs are off the queue."""
+        pass
+
+
+class DownloadQueueService(DownloadQueueServiceBase):
+    """Multithreaded queue for downloading models via URL or repo_id."""
+
+    _event_bus: EventServiceBase
+    _queue: DownloadQueue
+
+    def __init__(self, event_bus: EventServiceBase, **kwargs):
+        """
+        Initialize new DownloadQueueService object.
+
+        :param event_bus: EventServiceBase object for reporting progress.
+        :param **kwargs: Any of the arguments taken by invokeai.backend.model_manager.download.DownloadQueue.
+        e.g. `max_parallel_dl`.
+        """
+        self._event_bus = event_bus
+        self._queue = DownloadQueue()
+
+    def create_download_job(
+        self,
+        source: Union[str, Path, AnyHttpUrl],
+        destdir: Path,
+        filename: Optional[Path] = None,
+        start: bool = True,
+        access_token: Optional[str] = None,
+        event_handlers: Optional[List[DownloadEventHandler]] = None,
+    ) -> DownloadJobBase:  # noqa D102
+        event_handlers = event_handlers or []
+        if self._event_bus:
+            event_handlers.append([self._event_bus.emit_model_download_event])
```

Comment:
Method doesn't exist on events service

Reply:
This looks like old code and has other problems as well. Thanks for pointing out.